File: Python version/languages/asp/answer_sets.py
from .answer_set import AnswerSet
from abc import ABCMeta
from base.output import Output
from parsers.asp.asp_data_collection import ASPDataCollection
import sys


class AnswerSets(Output, ASPDataCollection):
    """A collection of AnswerSet."""
    __metaclass__ = ABCMeta

    def __init__(self, out, err=None):
        super(AnswerSets, self).__init__(out, err)
        self._answer_sets = None  # Where Answer Sets are stored

    def get_answer_sets(self):
        """Return a set of AnswerSet."""
        if self._answer_sets is None:
            self._answer_sets = list()
            self._parse()
        return self._answer_sets

    def get_optimal_answer_sets(self):
        answer_sets = self.get_answer_sets()
        levels = 0
        optimal_answer_sets = []

        for answer_set in answer_sets:
            max_level = max(answer_set.get_weights().keys())

            if levels < max_level:
                levels = max_level

        for level in range(levels, 0, -1):
            minimum_cost = sys.maxsize

            for answer_set in answer_sets:
                cost = answer_set.get_weights().get(level, 0)

                if cost < minimum_cost:
                    optimal_answer_sets.clear()

                    minimum_cost = cost

                if cost == minimum_cost:
                    optimal_answer_sets.append(answer_set)

            answer_sets = list(optimal_answer_sets)

        return optimal_answer_sets

    def get_answer_sets_string(self):
        """Return a string which contains every AnswerSet."""
        return self._output

    def add_answer_set(self):
        self._answer_sets.append(AnswerSet([], {}))

    def store_atom(self, result):
        self._answer_sets[-1].get_answer_set().append(result)

    def store_cost(self, level, weight):
        self._answer_sets[-1].get_weights()[int(level)] = int(weight)

File: Java version/app/src/main/java/it/unical/mat/embasp/languages/asp/AnswerSets.java
package it.unical.mat.embasp.languages.asp;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import it.unical.mat.embasp.base.Output;
import it.unical.mat.parsers.asp.ASPDataCollection;

/** A collection of {@link AnswerSet} */

public abstract class AnswerSets extends Output implements ASPDataCollection {
	/** Where Answer Sets are stored */
	protected List<AnswerSet> answersets;

	public AnswerSets(final String out) {
		super(out);
	}

	public AnswerSets(final String out, final String err) {
		super(out, err);
	}

	@Override
	public Object clone() throws CloneNotSupportedException {
		return super.clone();
	}

	public List<AnswerSet> getAnswersets() {
		if (answersets == null) {
			answersets = new ArrayList<>();
			parse();
		}

		return Collections.unmodifiableList(answersets);
	}

	public List<AnswerSet> getOptimalAnswerSets() {
		int level = 0;
		List <AnswerSet> answerSets = getAnswersets(), optimalAnswerSets = new ArrayList <> ();
		
		for(final AnswerSet answerSet : answerSets) {
			final int maxLevel = Collections.max(answerSet.getWeights().keySet());
			
			if(level < maxLevel)
				level = maxLevel;
		}
		
		for(; level >= 1; level--) {
			int minimumCost = Integer.MAX_VALUE;
			
			for(final AnswerSet answerSet : answerSets) {
				final int cost = answerSet.getWeights().getOrDefault(level, 0);
				
				if(cost < minimumCost) {
					optimalAnswerSets.clear();
					
					minimumCost = cost;
				}
				
				if(cost == minimumCost)
					optimalAnswerSets.add(answerSet);
			}

			answerSets = new ArrayList <> (optimalAnswerSets);
		}
		
		return optimalAnswerSets;
	}
	
	public String getAnswerSetsString() {
		return output;
	}

	@Override
	public void addAnswerSet() {
		answersets.add(new AnswerSet(new ArrayList <String> ()));
	}
	
	@Override
	public void storeAtom(final String result) {
		answersets.get(answersets.size() - 1).getValue().add(result);
	}
	
	@Override
	public void storeCost(final int level, final int weight) {
		answersets.get(answersets.size() - 1).getLevelWeight().put(level, weight);
	}
}
